File: python/packages/agent_runtime/src/insightgraph_agent/retriever_agent.py
from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


class RetrieverAgent:
    """Iterative retriever that follows graph connections across rounds.

    Unlike a simple for-loop executor, this agent:
    1. Executes the initial plan
    2. Extracts entities from results
    3. Generates follow-up queries to explore entity neighborhoods
    4. Stops when no new information is found or max_iterations reached
    """

    def __init__(self, tools: Any, max_iterations: int = 3):
        self._tools = tools
        self._max_iterations = max_iterations

    async def execute_plan(self, tool_plan: list[dict]) -> list[dict]:
        """Execute a tool plan with iterative graph exploration.

        Returns a list of {tool, args, result, error} dicts.
        """
        all_results: list[dict] = []
        discovered_entities: set[str] = set()

        # Round 1: Execute the original plan
        round_results = await self._execute_steps(tool_plan)
        all_results.extend(round_results)

        # Extract entities from results
        new_entities = self._extract_entities(round_results)
        discovered_entities.update(new_entities)

        # Round 2+: Follow graph connections
        for iteration in range(self._max_iterations - 1):
            follow_ups = self._generate_follow_ups(round_results, discovered_entities)
            if not follow_ups:
                logger.debug("No follow-up queries at iteration %d", iteration + 1)
                break

            logger.info(
                "Iteration %d: %d follow-up queries",
                iteration + 1,
                len(follow_ups),
            )
            round_results = await self._execute_steps(follow_ups)
            all_results.extend(round_results)

            new = self._extract_entities(round_results) - discovered_entities
            if not new:
                logger.debug("No new entities at iteration %d", iteration + 1)
                break
            discovered_entities.update(new)

        return all_results

    async def _execute_steps(self, steps: list[dict]) -> list[dict]:
        """Execute a list of tool steps."""
        results = []
        for step in steps:
            tool_name = step.get("tool", "")
            args = step.get("args", {})
            try:
                result = await self._tools.execute(tool_name, args)
                results.append(
                    {
                        "tool": tool_name,
                        "args": args,
                        "result": result,
                        "error": None,
                    }
                )
            except Exception as e:
                logger.warning("Tool %s failed: %s", tool_name, e, exc_info=True)
                results.append(
                    {
                        "tool": tool_name,
                        "args": args,
                        "result": None,
                        "error": str(e),
                    }
                )
        return results

    @staticmethod
    def _extract_entities(results: list[dict]) -> set[str]:
        """Extract entity names from tool execution results."""
        entities: set[str] = set()
        for r in results:
            result = r.get("result")
            if result is None:
                continue

            if isinstance(result, list):
                for item in result:
                    if isinstance(item, dict):
                        for key in ("name", "canonical_name", "entity_name", "mentioned_entity"):
                            val = item.get(key)
                            if val:
                                entities.add(val)
                        # Also check nested entity dicts
                        entity = item.get("entity")
                        if isinstance(entity, dict):
                            for key in ("name", "canonical_name"):
                                val = entity.get(key)
                                if val:
                                    entities.add(val)

            elif isinstance(result, dict):
                for key in ("name", "canonical_name"):
                    val = result.get(key)
                    if val:
                        entities.add(val)
                # Check nodes in subgraph results
                for node in result.get("nodes", []):
                    if isinstance(node, dict):
                        props = node.get("props", node)
                        for key in ("name", "canonical_name"):
                            val = props.get(key)
                            if val:
                                entities.add(val)

        return entities

    @staticmethod
    def _generate_follow_ups(
        results: list[dict],
        known_entities: set[str],
    ) -> list[dict]:
        """Generate follow-up tool calls based on discovered entities.

        Explores entity neighborhoods by querying for:
        - Claims about newly discovered entities
        - Relationships of entities
        - Evidence for high-confidence claims
        """
        follow_ups: list[dict] = []
        seen_queries: set[str] = set()

        for r in results:
            result = r.get("result")
            if result is None:
                continue

            # If we found entities, get claims about them
            entities = RetrieverAgent._extract_entities([r])
            for ename in entities:
                query_key = f"claims:{ename}"
                if query_key not in seen_queries and r.get("tool") != "get_claims_about":
                    follow_ups.append(
                        {
                            "tool": "get_claims_about",
                            "args": {"entity_name": ename},
                        }
                    )
                    seen_queries.add(query_key)

                # Also explore entity relationships
                rel_key = f"relationships:{ename}"
                if rel_key not in seen_queries:
                    follow_ups.append(
                        {
                            "tool": "find_related_entities",
                            "args": {"entity_name": ename, "depth": 1},
                        }
                    )
                    seen_queries.add(rel_key)

            # If we found claims, trace evidence for high-confidence ones
            if isinstance(result, list):
                for item in result:
                    if isinstance(item, dict):
                        claim_id = item.get("claim_id")
                        confidence = item.get("confidence", 0)
                        if claim_id and confidence and confidence > 0.7:
                            ev_key = f"evidence:{claim_id}"
                            if ev_key not in seen_queries:
                                follow_ups.append(
                                    {
                                        "tool": "find_evidence_for_claim",
                                        "args": {"claim_id": claim_id},
                                    }
                                )
                                seen_queries.add(ev_key)

        # Limit follow-ups per round
        return follow_ups[:10]

File: packages/agent-runtime/src/retriever-agent.ts
/**
 * Tool executor interface. The consumer must provide an object with an
 * `execute(toolName, args)` method that resolves to a result value.
 */
export interface ToolExecutor {
  execute(toolName: string, args: Record<string, unknown>): Promise<unknown>;
}

export interface StepResult {
  tool: string;
  args: Record<string, unknown>;
  result: unknown;
  error: string | null;
}

/**
 * Iterative retriever that follows graph connections across rounds.
 *
 * Unlike a simple for-loop executor, this agent:
 * 1. Executes the initial plan
 * 2. Extracts entities from results
 * 3. Generates follow-up queries to explore entity neighborhoods
 * 4. Stops when no new information is found or max_iterations reached
 */
export class RetrieverAgent {
  private tools: ToolExecutor;
  private maxIterations: number;

  constructor(tools: ToolExecutor, maxIterations = 3) {
    this.tools = tools;
    this.maxIterations = maxIterations;
  }

  /**
   * Execute a tool plan with iterative graph exploration.
   * Returns a list of step result dicts.
   */
  async executePlan(
    toolPlan: Array<{ tool: string; args: Record<string, unknown> }>,
  ): Promise<StepResult[]> {
    const allResults: StepResult[] = [];
    const discoveredEntities = new Set<string>();

    // Round 1: Execute the original plan
    let roundResults = await this.executeSteps(toolPlan);
    allResults.push(...roundResults);

    // Extract entities from results
    const newEntities = RetrieverAgent.extractEntities(roundResults);
    for (const e of newEntities) discoveredEntities.add(e);

    // Round 2+: Follow graph connections
    for (let iteration = 0; iteration < this.maxIterations - 1; iteration++) {
      const followUps = RetrieverAgent.generateFollowUps(roundResults, discoveredEntities);
      if (followUps.length === 0) break;

      roundResults = await this.executeSteps(followUps);
      allResults.push(...roundResults);

      const freshEntities = RetrieverAgent.extractEntities(roundResults);
      let hasNew = false;
      for (const e of freshEntities) {
        if (!discoveredEntities.has(e)) {
          discoveredEntities.add(e);
          hasNew = true;
        }
      }
      if (!hasNew) break;
    }

    return allResults;
  }

  private async executeSteps(
    steps: Array<{ tool: string; args: Record<string, unknown> }>,
  ): Promise<StepResult[]> {
    const results: StepResult[] = [];
    for (const step of steps) {
      const toolName = step.tool ?? "";
      const args = step.args ?? {};
      try {
        const result = await this.tools.execute(toolName, args);
        results.push({ tool: toolName, args, result, error: null });
      } catch (e: unknown) {
        const message = e instanceof Error ? e.message : String(e);
        console.warn(`Tool ${toolName} failed: ${message}`);
        results.push({ tool: toolName, args, result: null, error: message });
      }
    }
    return results;
  }

  /**
   * Extract entity names from tool execution results.
   */
  static extractEntities(results: StepResult[]): Set<string> {
    const entities = new Set<string>();

    for (const r of results) {
      const result = r.result;
      if (result == null) continue;

      if (Array.isArray(result)) {
        for (const item of result) {
          if (item && typeof item === "object") {
            const rec = item as Record<string, unknown>;
            for (const key of ["name", "canonical_name", "entity_name", "mentioned_entity"]) {
              const val = rec[key];
              if (typeof val === "string" && val) entities.add(val);
            }
            // Also check nested entity dicts
            const entity = rec.entity;
            if (entity && typeof entity === "object") {
              const entityRec = entity as Record<string, unknown>;
              for (const key of ["name", "canonical_name"]) {
                const val = entityRec[key];
                if (typeof val === "string" && val) entities.add(val);
              }
            }
          }
        }
      } else if (typeof result === "object") {
        const rec = result as Record<string, unknown>;
        for (const key of ["name", "canonical_name"]) {
          const val = rec[key];
          if (typeof val === "string" && val) entities.add(val);
        }
        // Check nodes in subgraph results
        const nodes = rec.nodes;
        if (Array.isArray(nodes)) {
          for (const node of nodes) {
            if (node && typeof node === "object") {
              const nodeRec = node as Record<string, unknown>;
              const props = (nodeRec.props ?? nodeRec) as Record<string, unknown>;
              for (const key of ["name", "canonical_name"]) {
                const val = props[key];
                if (typeof val === "string" && val) entities.add(val);
              }
            }
          }
        }
      }
    }

    return entities;
  }

  /**
   * Generate follow-up tool calls based on discovered entities.
   *
   * Explores entity neighborhoods by querying for:
   * - Claims about newly discovered entities
   * - Relationships of entities
   * - Evidence for high-confidence claims
   */
  static generateFollowUps(
    results: StepResult[],
    knownEntities: Set<string>,
  ): Array<{ tool: string; args: Record<string, unknown> }> {
    const followUps: Array<{ tool: string; args: Record<string, unknown> }> = [];
    const seenQueries = new Set<string>();

    for (const r of results) {
      const result = r.result;
      if (result == null) continue;

      // If we found entities, get claims about them
      const entities = RetrieverAgent.extractEntities([r]);
      for (const ename of entities) {
        const queryKey = `claims:${ename}`;
        if (!seenQueries.has(queryKey) && r.tool !== "get_claims_about") {
          followUps.push({
            tool: "get_claims_about",
            args: { entity_name: ename },
          });
          seenQueries.add(queryKey);
        }

        // Also explore entity relationships
        const relKey = `relationships:${ename}`;
        if (!seenQueries.has(relKey)) {
          followUps.push({
            tool: "find_related_entities",
            args: { entity_name: ename, depth: 1 },
          });
          seenQueries.add(relKey);
        }
      }

      // If we found claims, trace evidence for high-confidence ones
      if (Array.isArray(result)) {
        for (const item of result) {
          if (item && typeof item === "object") {
            const rec = item as Record<string, unknown>;
            const claimId = rec.claim_id;
            const confidence = typeof rec.confidence === "number" ? rec.confidence : 0;
            if (claimId && confidence > 0.7) {
              const evKey = `evidence:${claimId}`;
              if (!seenQueries.has(evKey)) {
                followUps.push({
                  tool: "find_evidence_for_claim",
                  args: { claim_id: claimId },
                });
                seenQueries.add(evKey);
              }
            }
          }
        }
      }
    }

    // Limit follow-ups per round
    return followUps.slice(0, 10);
  }
}
